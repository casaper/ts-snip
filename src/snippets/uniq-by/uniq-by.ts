/**
 * Unique an array of records by key
 *
 * [Playground Link](https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&noImplicitReturns=false&target=7&jsx=0&module=6&pretty=true&allowSyntheticDefaultImports=false#code/GYVwdgxgLglg9mABOGBHAQgTwDwChGIAqiApgB5QlgAmAzogEokRwBO12tUrMYA5ogA+iMCAC2AIxKshiWpklwANgBpkYANZg4AdzAA+FfkQBpUhSp1EGkpjjAiufQAoWSpc1gIAXEQDaALpqNpi+JgCUvoSBiADexqwkUCCsSG4e0PBgxgQAdMAwSpSszjCUYogAvPrWtoi89eXhOYj5hcXOLQSucGIADgCGiQBirL1qvNTkaiioICTUAIKsrAOY4VX6XQT1NORVlZXbBLPzSytrbTQAkntkPf1DJIRwm4gsjyNjYn4hAQeVd69QaJF6-WwBZo7RBQna5FiQAZQHruTxZNpFaSlcpvACEzhCu0aJDE4TJAG5cABfXC4BFcOQDfoeeiAvzGWKIAa+ACMagkvMQVKMBE53MQACZ+YLhRyubzpYgeUKRXF5UrFXzENQZarORBfFLtbq5QaNcalSrcAFafSoFyquo0FhnLQmX0WWoAOQDL3hSl2xASR2zF1u5kkWjeiR+gMIBkQENgZ2YV3uz2IL0QWN0+P26hJlNpiNRzPUHO5sAMxK0EBFVlqgb8tQQNQFqmUyu0ZQkXJKOB8Zw1utQWj+3BAA)
 */
export function uniqBy<
  T extends Record<string | number | symbol, unknown>,
  K extends keyof T,
>(collection: T[], key: K): T[] {
  return collection
    .filter(item => key in item)
    .filter(
      (compareFrom, index, uniqArray) =>
        index ===
        uniqArray.findIndex(compareTo => {
          return compareFrom[key] === compareTo[key];
        }),
    )
    .concat(collection.filter(item => !(key in item)));
}
