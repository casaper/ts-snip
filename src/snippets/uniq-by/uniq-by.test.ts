import { uniqBy } from "./uniq-by";

fdescribe('uniqBy', () => {
  const samples = [
    { a: 1, b: 1 },
    { a: 2, b: 1 },
    { a: 1, b: 1 },
    { a: 1, b: 1, d: 1 },
    { c: 2, d: 1 },
    { c: 1, d: 1 },
  ];

  test('uniqBy "a"', () => {
    const result = uniqBy(samples, 'a');
    expect(result).toHaveLength(4);
    expect(result).toMatchSnapshot();
  });

  test('uniqBy "b"', () => {
    const result = uniqBy(samples, 'b');
    expect(result).toHaveLength(3);
    expect(result).toMatchSnapshot();
  });
});
