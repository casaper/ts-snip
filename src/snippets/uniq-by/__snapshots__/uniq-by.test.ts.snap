// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uniqBy uniqBy "a" 1`] = `
Array [
  Object {
    "a": 1,
    "b": 1,
  },
  Object {
    "a": 2,
    "b": 1,
  },
  Object {
    "c": 2,
    "d": 1,
  },
  Object {
    "c": 1,
    "d": 1,
  },
]
`;

exports[`uniqBy uniqBy "b" 1`] = `
Array [
  Object {
    "a": 1,
    "b": 1,
  },
  Object {
    "c": 2,
    "d": 1,
  },
  Object {
    "c": 1,
    "d": 1,
  },
]
`;
